### ISSUE TITLE
FailedScheduling didn't have free ports

### ISSUE THREAD

**Issue by username1366** [kubernetes/kubernetes#66568]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
After deploy I have some pods in Pending state and they won't change his state untill I restart kube-scheduler. 
kubectl describe pod \<pod name\>:
```
Warning  FailedScheduling  20m (x745 over 24m)   default-scheduler  0/72 nodes are available: 23 node(s) didn't match node selector, 43 node(s) didn't have free ports for the requested pod ports, 45 Insufficient cpu.
```

**What you expected to happen**:
All pods should be in Running state after deployment

**How to reproduce it (as minimally and precisely as possible)**:

```
kubectl apply -f test-deploy.yaml
kubectl apply -f test-deploy2.yaml
```

test-deploy.yaml
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-deploy
spec:
  replicas: 25
  selector:
    matchLabels:
      deployment: test-deploy
  template:
    metadata:
      labels:
        app: test-deploy
        deployment: test-deploy
    spec:
      hostNetwork: true
      dnsPolicy: ClusterFirstWithHostNet
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: dedicated
                operator: In
                values: [ 'api-test' ]
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values: [ 'test-deploy']
            topologyKey: kubernetes.io/hostname
      tolerations:
      - key: dedicated
        operator: Equal
        value: api-test
        effect: NoExecute
      containers:
      - name: envoy-prometheus
        image: example/envoy-exporter:latest
        env:
        - name: SERVER_PORT
          value: "9092"
        - name: URL
          value: "http://127.0.0.1/stats"
        ports:
        - name: prom-stats
          containerPort: 9092
      imagePullSecrets:
      - name: dockerhubkey
```

test-deploy2.yaml
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test-deploy2
spec:
  replicas: 16
  selector:
    matchLabels:
      deployment: test-deploy2
  template:
    metadata:
      labels:
        app: test-deploy
        deployment: test-deploy2
    spec:
      hostNetwork: true
      dnsPolicy: ClusterFirstWithHostNet
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: kubernetes.io/hostname
                operator: In
                values:
                - 'app-pool2-1-europe-central1.testdc.com'
                - 'app-pool2-2-europe-central1.testdc.com'
                - 'app-pool2-3-europe-central1.testdc.com'
                - 'app-pool2-4-europe-central1.testdc.com'
                - 'app-pool2-5-europe-central1.testdc.com'
                - 'app-pool2-6-europe-central1.testdc.com'
                - 'app-pool2-7-europe-central1.testdc.com'
                - 'app-pool2-8-europe-central1.testdc.com'
                - 'app-pool2-9-europe-central1.testdc.com'
                - 'app-pool2-10-europe-central1.testdc.com'
                - 'app-pool2-11-europe-central1.testdc.com'
                - 'app-pool2-12-europe-central1.testdc.com'
                - 'app-pool2-13-europe-central1.testdc.com'
                - 'app-pool2-14-europe-central1.testdc.com'
                - 'app-pool2-15-europe-central1.testdc.com'
                - 'app-pool2-16-europe-central1.testdc.com'
                - 'app-pool2-17-europe-central1.testdc.com'
                - 'app-pool2-18-europe-central1.testdc.com'
                - 'app-pool2-19-europe-central1.testdc.com'
                - 'app-pool2-20-europe-central1.testdc.com'
                - 'app-pool2-21-europe-central1.testdc.com'
                - 'app-pool2-22-europe-central1.testdc.com'
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values: [ 'test-deploy' ]
            topologyKey: kubernetes.io/hostname
      tolerations:
      - key: dedicated
        operator: Equal
        value: api-test
        effect: NoExecute
      containers:
      - name: envoy-prometheus
        image: test/envoy-exporter:latest
        env:
        - name: SERVER_PORT
          value: "9092"
        - name: URL
          value: "http://127.0.0.1/stats"
        ports:
        - name: prom-stats
          containerPort: 9092
      imagePullSecrets:
      - name: dockerhubkey
```

Describe Pending pod
```
kubectl describe pod test-deploy2-675ccc77b7-2gmsm
 Warning  FailedScheduling  59s   default-scheduler     0/72 nodes are available: 47 node(s) didn't have free ports for the requested pod ports, 54 node(s) didn't match node selector.
```

Restart kube-scheduler container
```
crictl stop $(crictl ps | grep kube-scheduler | awk '{print $1}')
```

Describe Pending pod
```
kubectl describe pod test-deploy2-675ccc77b7-2gmsm
  Normal   Scheduled         1m   default-scheduler     Successfully assigned default/test-deploy2-675ccc77b7-2gmsm to app-pool2-3-europe-central1.testdc.com
```

**Anything else we need to know?**:

I have two different deployments which use same Host ports but rise up on different nodes. It worked good when I used kubernetes v1.10 and it was broken when I updated to v.1.11

**Environment**:
- Kubernetes version (use `kubectl version`): v1.11.0
- Cloud provider or hardware configuration: self-hosted AMD EPYC 7401P 24-Core 128 GB DDR4
- OS (e.g. from /etc/os-release): Ubuntu 18.04 LTS
- Kernel (e.g. `uname -a`): 4.15.0-24-generic
- Install tools: kubeadm
- Others: containerd v1.12, calico network v3.1

**Comment by username1366** [comment]
Update
It also happens when I use one deploymend for all nodes and hostNetwork: false

```
test-deploy-74c7d7db9d-q2x2b                      1/1       Running     0          24m       10.29.129.65      app-pool2-11-europe-central1.testdc.com
test-deploy-7cbbf48ccd-25ps7                      1/1       Running     0          15m       10.18.176.194     app12-europe-central1.testdc.com
test-deploy-7cbbf48ccd-2clbr                      1/1       Running     0          13m       10.12.178.66      app15.europe-central1.testdc.com
test-deploy-7cbbf48ccd-4fnnl                      0/1       Pending     0          11m       <none>            <none>    app21.europe-central1.testdc.com
test-deploy-7cbbf48ccd-4p59b                      1/1       Running     0          14m       10.43.72.2        app20.europe-central1.testdc.com
test-deploy-7cbbf48ccd-4xqbk                      0/1       Pending     0          13m       <none>            <none>
test-deploy-7cbbf48ccd-5hj9m                      0/1       Pending     0          14m       <none>            <none>
test-deploy-7cbbf48ccd-6bhs2                      1/1       Running     0          15m       10.47.112.66      app18.europe-central1.testdc.com
test-deploy-7cbbf48ccd-6ntq7                      1/1       Running     0          15m       10.55.159.2       app5.europe-central1.testdc.com
test-deploy-7cbbf48ccd-7f2hr                      0/1       Pending     0          15m       <none>            <none>
test-deploy-7cbbf48ccd-7qm95                      1/1       Running     0          15m       10.12.29.193      app-pool2-1-europe-central1.testdc.com
test-deploy-7cbbf48ccd-975t6                      1/1       Running     0          14m       10.7.162.194      app-pool2-9-europe-central1.testdc.com
test-deploy-7cbbf48ccd-9kdkp                      1/1       Running     0          15m       10.1.149.66       app27.europe-central1.testdc.com
test-deploy-7cbbf48ccd-9xtwn                      0/1       Pending     0          15m       <none>            <none>    app22.europe-central1.testdc.com
test-deploy-7cbbf48ccd-b4zpt                      1/1       Running     0          15m       10.47.158.130     app2.europe-central1.testdc.com
test-deploy-7cbbf48ccd-ctllc                      1/1       Running     0          15m       10.53.181.66      app13.europe-central1.testdc.com
test-deploy-7cbbf48ccd-d88cx                      0/1       Pending     0          12m       <none>            <none>
test-deploy-7cbbf48ccd-f4wvx                      0/1       Pending     0          15m       <none>            <none>
test-deploy-7cbbf48ccd-fh6f7                      1/1       Running     0          15m       10.56.219.2       app6.europe-central1.testdc.com
test-deploy-7cbbf48ccd-fnb2g                      0/1       Pending     0          14m       <none>            <none>
test-deploy-7cbbf48ccd-g4ssq                      1/1       Running     0          14m       10.54.240.194     app-pool2-19-europe-central1.testdc.com
test-deploy-7cbbf48ccd-g98kf                      1/1       Running     0          15m       10.12.181.66      app-pool2-14-europe-central1.testdc.com
test-deploy-7cbbf48ccd-gdcz4                      1/1       Running     0          15m       10.51.122.66      app8.europe-central1.testdc.com
test-deploy-7cbbf48ccd-gjb6x                      1/1       Running     0          15m       10.61.170.130     app10.europe-central1.testdc.com
test-deploy-7cbbf48ccd-gwfsq                      1/1       Running     0          15m       10.39.167.130     app-pool2-10-europe-central1.testdc.com
test-deploy-7cbbf48ccd-gx6jq                      1/1       Running     0          15m       10.38.127.66      app-pool2-15-europe-central1.testdc.com
test-deploy-7cbbf48ccd-hgtzt                      1/1       Running     0          14m       10.44.235.66      app-pool2-4-europe-central1.testdc.com
test-deploy-7cbbf48ccd-hn4gw                      1/1       Running     0          15m       10.12.152.194     app16.europe-central1.testdc.com
test-deploy-7cbbf48ccd-hv9bv                      1/1       Running     0          14m       10.14.191.194     app19.europe-central1.testdc.com
test-deploy-7cbbf48ccd-hxxx2                      1/1       Running     0          13m       10.55.129.66      app-pool2-5-europe-central1.testdc.com
test-deploy-7cbbf48ccd-jfp9w                      1/1       Running     0          15m       10.32.118.66      app-pool2-3-europe-central1.testdc.com
test-deploy-7cbbf48ccd-jsw79                      1/1       Running     0          12m       10.47.110.130     app-pool2-22-europe-central1.testdc.com
test-deploy-7cbbf48ccd-kjjvg                      1/1       Running     0          15m       10.51.189.194     app-pool2-16-europe-central1.testdc.com
test-deploy-7cbbf48ccd-mdb5k                      0/1       Pending     0          11m       <none>            <none>
test-deploy-7cbbf48ccd-msb5s                      1/1       Running     0          13m       10.0.212.2        app11.europe-central1.testdc.com
test-deploy-7cbbf48ccd-nbwsr                      1/1       Running     0          13m       10.44.197.130     app-pool2-2-europe-central1.testdc.com
test-deploy-7cbbf48ccd-nn85q                      0/1       Pending     0          14m       <none>            <none>    app4.europe-central1.testdc.com
test-deploy-7cbbf48ccd-phtrg                      0/1       Pending     0          12m       <none>            <none>
test-deploy-7cbbf48ccd-qv6hs                      0/1       Pending     0          15m       <none>            <none>
test-deploy-7cbbf48ccd-rc77t                      1/1       Running     0          15m       10.62.187.129     app7.europe-central1.testdc.com
test-deploy-7cbbf48ccd-rm8kw                      1/1       Running     0          15m       10.18.23.2        app17.europe-central1.testdc.com
test-deploy-7cbbf48ccd-t4w7b                      1/1       Running     0          15m       10.32.48.2        app1.europe-central1.testdc.com
test-deploy-7cbbf48ccd-wbwmn                      1/1       Running     0          15m       10.19.148.2       app-pool2-21-europe-central1.testdc.com
test-deploy-7cbbf48ccd-wfjb9                      1/1       Running     0          15m       10.3.93.194       app-pool2-6-europe-central1.testdc.com
test-deploy-7cbbf48ccd-wtkpn                      1/1       Running     0          14m       10.13.233.66      app-pool2-20-europe-central1.testdc.com
test-deploy-7cbbf48ccd-xwh2n                      1/1       Running     0          15m       10.38.46.1        app-pool2-17-europe-central1.testdc.com
```

kube-scheduler logs
```
W0724 12:25:38.643079       1 options.go:148] WARNING: all flags other than --config, --write-config-to, and --cleanup are deprecated. Please begin using a config file ASAP.
I0724 12:25:38.646706       1 server.go:126] Version: v1.11.0
W0724 12:25:38.647630       1 authorization.go:47] Authorization is disabled
W0724 12:25:38.647647       1 authentication.go:55] Authentication is disabled
I0724 12:25:38.647657       1 insecure_serving.go:47] Serving healthz insecurely on 127.0.0.1:10251
I0724 12:25:39.548643       1 controller_utils.go:1025] Waiting for caches to sync for scheduler controller
I0724 12:25:39.648762       1 controller_utils.go:1032] Caches are synced for scheduler controller
I0724 12:25:39.648794       1 leaderelection.go:185] attempting to acquire leader lease  kube-system/kube-scheduler...
I0724 12:25:57.151617       1 leaderelection.go:194] successfully acquired lease kube-system/kube-scheduler
```

**Comment by username1366** [comment]
I've built kube-scheduler with some debug info output https://github.com/username1366/kubernetes/pull/1/files It prints nodeInfo.UsedPorts() and schedutil.GetContainerPorts(pod) when portsConflict happens. 

Example debug output:
```
nodeInfo: app-pool2-13-europe-central1.testdc.com ================  
existingPorts: map[{Protocol:TCP Port:443}:{} {Protocol:TCP Port:8001}:{} {Protocol:TCP Port:9091}:{} {Protocol:UDP Port:8125}:{} {Protocol:TCP Port:19999}:{} {Protocol:UDP Port:8126}:{} {Protocol:TCP Port:8080}:{} {Protocol:TCP Port:80}:{}] ================ 
wantPorts: [&ContainerPort{Name:,HostPort:8080,ContainerPort:8080,Protocol:TCP,HostIP:,} &ContainerPort{Name:http,HostPort:80,ContainerPort:80,Protocol:TCP,HostIP:,} &ContainerPort{Name:https,HostPort:443,ContainerPort:443,Protocol:TCP,HostIP:,} &ContainerPort{Name:stats,HostPort:8001,ContainerPort:8001,Protocol:TCP,HostIP:,} &ContainerPort{Name:prom-stats,HostPort:9091,ContainerPort:9091,Protocol:TCP,HostIP:,}] ================ 
podName: test-deploy-dccc46c97-dz286
```

If we look to debug output we'll see that kube-scheduler trying to schedule pod test-deploy-dccc46c97-dz286 to node app-pool2-13-europe-central1.testdc.com and it couldn't do that because of portsConflict (container port: {HostPort:443,ContainerPort:443}, node used port: {Protocol:TCP Port:443}) But if we look into app-pool2-13-europe-central1.testdc.com node we find out that nobody use 443 tcp port

```
root@app-pool2-13-europe-central1 /home/ubuntu # ss -nltp
State           Recv-Q           Send-Q                      Local Address:Port                      Peer Address:Port
LISTEN          0                65534                           127.0.0.1:5000                           0.0.0.0:*              users:(("agent",pid=27424,fd=3))
LISTEN          0                65534                           127.0.0.1:10248                          0.0.0.0:*              users:(("kubelet",pid=23732,fd=17))
LISTEN          0                65534                           127.0.0.1:10249                          0.0.0.0:*              users:(("kube-proxy",pid=27625,fd=7))
LISTEN          0                65534                           127.0.0.1:5001                           0.0.0.0:*              users:(("agent",pid=27424,fd=6))
LISTEN          0                8                                 0.0.0.0:179                            0.0.0.0:*              users:(("bird",pid=26495,fd=7))
LISTEN          0                128                         127.0.0.53%lo:53                             0.0.0.0:*              users:(("systemd-resolve",pid=795,fd=13))
LISTEN          0                128                               0.0.0.0:22                             0.0.0.0:*              users:(("sshd",pid=1129,fd=3))
LISTEN          0                65534                           127.0.0.1:10010                          0.0.0.0:*              users:(("containerd",pid=14815,fd=62))
LISTEN          0                65534                           127.0.0.1:8126                           0.0.0.0:*              users:(("trace-agent",pid=27426,fd=5))
LISTEN          0                4096                              0.0.0.0:19999                          0.0.0.0:*              users:(("netdata",pid=7500,fd=3))
LISTEN          0                65534                                   *:10250                                *:*              users:(("kubelet",pid=23732,fd=16))
LISTEN          0                65534                                   *:9099                                 *:*              users:(("calico-felix",pid=26496,fd=3))
LISTEN          0                65534                                   *:10256                                *:*              users:(("kube-proxy",pid=27625,fd=6))
LISTEN          0                128                                  [::]:22                                [::]:*              users:(("sshd",pid=1129,fd=4))
```

I guess something wrong with kube-scheduler cache or my configuration. @k82cn could you help please?

**Comment by jpiper** [comment]
I am also having this issue since upgrading from 1.10 to 1.11.

When I have `hostNetwork:true` then specifying `containerPort` prevents the pod from being scheduled.

**Comment by jpiper** [comment]
I think this commit broke it:

https://github.com/kubernetes/kubernetes/commit/68c2c79362a506939b6513108b9b8d689cf94b0f#diff-5e800826dd0c3c6ec24dc167d287aea2

**Comment by jpiper** [comment]
@anfernee are you able to take a look?

**Comment by jpiper** [comment]
Yep, I can see via `lsof` that the port isn't in use in my case either

**Comment by k82cn** [comment]
@Huang-Wei , please help to check this one; I can help on reviewing :)

**Comment by Huang-Wei** [comment]
sure

/assign

**Comment by Huang-Wei** [comment]
@username1366 I'm not able to reproduce the issue. Can I clarify your steps a little bit:

To simplify your case, let's say we got a **X**-node cluster (in your case X = 72), all labeled "dedicated:api-test".

Then in the 1st deployment yaml, it tries to deploy a **Y**-replica app (in your case Y = 25) - it has podAntiAffinity and nodeAffinity (with condition key dedicated in api-test).

In the 2nd deployment yaml, it tries to deploy a **Z**-replica app (in your case the Z = 16) - it also has podAntiAffinity and nodeAffinity (with condition key kubernetes.io/hostname in [host-1 ... host-z]).

X > Y > Z.

So here are my questions:

- do you always create 1st deployment prior to 2nd one?
- is X running fine all the time?
    - if yes, could you check the Pending pods of 2nd deployment. Are they *all* scheduled onto the nodes which 1st deployment occupies?
- can you try to create 2nd deployment first (waiting to see if all pods are running), and then 1st?

@jpiper are you using the same procedure? Simplified or different one could be helpful.

**Comment by jpiper** [comment]
@Huang-Wei I'm using a deployment with a `spec` like this

```yaml
      hostNetwork: true
      dnsPolicy: ClusterFirstWithHostNet
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: kubernetes.io/hostname
                operator: In
                values:
                - host1.corp.com
                - host2.corp.com
```

and then in one of the containers of the pod,

```yaml
        ports:
        - containerPort: 9123
          hostPort: 9123
          name: metrics
```

I will try and see if I can replicate on a minikube instance tomorrow

**Comment by Huang-Wei** [comment]
@username1366 @jpiper when you hit the issue, are there any pods stay in "Terminating" status? If so, those ones are actually holding the internal hostPorts scheduling cache.

**Comment by jpiper** [comment]
@Huang-Wei Nope - 100% the pod was terminated.

I even tried rebooting the node the pod was supposed to be scheduled on and it was still stuck in pending as no ports available. I didn't try restarting the scheduler though.

As soon as I downgraded the master components and kubelets to 1.10.2 all the pods were successfully scheduled.

I have a small cluster I can test this with next week if you have more info or tests/logs/debug info you'd like me to run

**Comment by Huang-Wei** [comment]
@username1366 @jpiper also want to highlight that when you hit the issue, you don't need to go to the physical node to check if the port is occupied (netstat/ss/etc.). Actually scheduler do not check real-time ports occupation, instead, it relies on the info from all running pods (those ones which uses `hostNetwork` and opened `containerPort`) and make scheduling decisions.

@username1366 from the log you provided, it seems the pod (failed to schedule) tried to claim ports 8080, 80, 443, 8001 and 9091 (from wantPorts info), and unfortunately some of them have been **claimed to** occupied - from existingPorts info:

```
existingPorts: 
map[
{Protocol:TCP Port:443}:{} 
{Protocol:TCP Port:8001}:{} 
{Protocol:TCP Port:9091}:{} 
{Protocol:UDP Port:8125}:{} <== check which one uses this port
{Protocol:TCP Port:19999}:{} <== check which one uses this port
{Protocol:UDP Port:8126}:{} <== check which one uses this port
{Protocol:TCP Port:8080}:{} 
{Protocol:TCP Port:80}:{}
]
```

So could you help to check which workloads have been using 8125, 19999, 8126. I know it's irrelevant, but I'm thinking from these ports, it's helpful to check if those workloads **claims** one of 8080, 80, 443, 8001 and 9091.

**Comment by Huang-Wei** [comment]
@jpiper thanks for the feedback. When you his the issue again, could you:

1. run the following command:

    > kubectl get po --all-namespaces -o=jsonpath="{range .items[*]}{.spec.nodeName}{'\t'}{.spec.hostNetwork}{'\t'}{.metadata.namespace}{'\t'}{.metadata.name}{'\t'}{.spec.hostNetwork}{'\t'}{.spec.containers..containerPort}{'\n'}{end}"

2. run `kubectl get po --all-namespaces -o wide | grep Pending` to list all Pending pods

**Comment by username1366** [comment]
@Huang-Wei sorry for my long absence
```
{Protocol:UDP Port:8125}:{} <== datadog statsd port controlled by daemonset
{Protocol:TCP Port:19999}:{} <== netdata web port controlled by daemonset
{Protocol:UDP Port:8126}:{} <== netdata statsd port controlled by daemonset
```
Actualy I don't have any problems with this pods they work fine. I hit an issue only after I update deployment via command:
```
kubectl apply --context europe-central1-new -f datacenters/tier1-europe-central1-new \
--recursive --prune -l kubernetes.io/cluster-service!=true \
--prune-whitelist=core/v1/ConfigMap --prune-whitelist=core/v1/Endpoints \
--prune-whitelist=core/v1/Pod --prune-whitelist=core/v1/ReplicationController \
--prune-whitelist=core/v1/Secret --prune-whitelist=core/v1/Service \
--prune-whitelist=batch/v1/Job --prune-whitelist=batch/v1beta1/CronJob \
--prune-whitelist=extensions/v1beta1/DaemonSet \
--prune-whitelist=extensions/v1beta1/Deployment \
--prune-whitelist=extensions/v1beta1/Ingress \
--prune-whitelist=extensions/v1beta1/ReplicaSet \
--prune-whitelist=apps/v1beta1/StatefulSet --prune-whitelist=apps/v1beta1/Deployment
```

And issue happens when old pods were terminated and nodes were free except netdata, datadog, and calico

> run the following command:
kubectl get po --all-namespaces -o=jsonpath="{range .items[*]}{.spec.nodeName}{'\t'}{.spec.hostNetwork}{'\t'}{.metadata.namespace}{'\t'}{.metadata.name}{'\t'}{.spec.hostNetwork}{'\t'}{.spec.containers..containerPort}{'\n'}{end}"

I will check it tomorrow I guess

Should I provide additional info?

**Comment by Huang-Wei** [comment]
@username1366 do you mean everytime you hit the issue when doing `kubectl apply -f`? I'm a little confused..

And could you answer the questions I proposed in comment https://github.com/kubernetes/kubernetes/issues/66568#issuecomment-410018330?

**Comment by Huang-Wei** [comment]
@jpiper could you show me result of `k get pod app-passive-7f97555865-4ntxx -o yaml -n app`? And you hit the issue on first attempt of deployment?

**Comment by jpiper** [comment]
@Huang-Wei when it was stuck pending it looked like this. I could downgrade and re-upgrade my cluster to break it again for you if you want more info!

```
% k describe pod app-passive-7f97555865-4ntxx 
Name:               app-passive-7f97555865-4ntxx
Namespace:          app
Node:               <none>
Labels:             app=app
                    pod-template-hash=3953111421
                    shard=passive
Annotations:        kubernetes.io/psp=privileged
Status:             Pending
IP:                 
Controlled By:      ReplicaSet/app-passive-7f97555865
NominatedNodeName:  0626.corp.com
Containers:
  app:
    Image:       app
    Ports:       8123/TCP, 9000/TCP, 9009/TCP
    Host Ports:  8123/TCP, 9000/TCP, 9009/TCP
...
  app-exporter:
    Image:      app
    Port:       9123/TCP
    Host Port:  9123/TCP

Conditions:
  Type           Status
  PodScheduled   False 
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason            Age                 From               Message
  ----     ------            ----                ----               -------
  Warning  FailedScheduling  11m (x14 over 11m)  default-scheduler  0/108 nodes are available: 2 node(s) were not ready, 2 node(s) were out of disk space, 29 Insufficient memory, 29 node(s) didn't have free ports for the requested pod ports, 3 node(s) were unschedulable, 91 node(s) didn't match node selector.
  Warning  FailedScheduling  11m (x7 over 11m)   default-scheduler  0/108 nodes are available: 2 node(s) were not ready, 2 node(s) were out of disk space, 29 Insufficient memory, 29 node(s) didn't have free ports for the requested pod ports, 3 node(s) were unschedulable, 92 node(s) didn't match node selector.
  Warning  FailedScheduling  11m (x3 over 11m)   default-scheduler  0/108 nodes are available: 2 node(s) were not ready, 2 node(s) were out of disk space, 29 Insufficient memory, 29 node(s) didn't have free ports for the requested pod ports, 3 node(s) were unschedulable, 93 node(s) didn't match node selector.
  Warning  FailedScheduling  11m                 default-scheduler  0/108 nodes are available: 2 node(s) were not ready, 2 node(s) were out of disk space, 29 Insufficient memory, 29 node(s) didn't have free ports for the requested pod ports, 3 node(s) were unschedulable, 94 node(s) didn't match node selector.
  Warning  FailedScheduling  6m (x275 over 7m)   default-scheduler  0/108 nodes are available: 2 node(s) were not ready, 2 node(s) were out of disk space, 28 Insufficient memory, 29 node(s) didn't have free ports for the requested pod ports, 3 node(s) were unschedulable, 91 node(s) didn't match node selector.
  Warning  FailedScheduling  1m (x499 over 7m)   default-scheduler  0/108 nodes are available: 2 node(s) were not ready, 2 node(s) were out of disk space, 28 Insufficient memory, 29 node(s) didn't have free ports for the requested pod ports, 3 node(s) were unschedulable, 93 node(s) didn't match node selector.
```

**Comment by jpiper** [comment]
@Huang-Wei note that in this case, the pod was scheduled before the upgrade, and then when I upgraded k8s and deleted the pod, the replacement one failed to schedule and gets stuck in pending. Restarting the control plane fixed this for me.

**Comment by Huang-Wei** [comment]
@jpiper unfortunately I still can't reproduce - even tried the upgrade procedure you mentioned.

I'm also curious how you do upgrade, kubeadm? shouldn't upgrade control plan auto restart scheduler?

(#65483 _might_ be related if you're using specific scheduler config.)

I'm a little stuck.. @k82cn any idea?

**Comment by Huang-Wei** [comment]
BTW: @jpiper output of `k get pod app-passive-7f97555865-4ntxx -o yaml -n app` still helps. Could you provide that? (k describe doesn't provide all the details)

**Comment by jpiper** [comment]
@Huang-Wei so in this case we are managing our k8s install via our own orchestration tool. Maybe this is the issue.

First, master components (scheduler, apiserver, controller-manager) are all upgraded (we are in a HA environment with 5 masters) and after that all the kubelets are restarted.

For the `k get pod` I will have to break my cluster again to get this for you :p so I will respond but within a few days maybe :)

**Comment by sfudeus** [comment]
Same issue here on a small cluster (3 master, 3 worker) with a metallb-speaker deployment with rack anti-affinity, hostNetwork, and node-selector for worker-nodes. 
Worked well with 1.10.5, does not now on 1.11.2.

```
   spec:
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchExpressions:
                - key: component
                  operator: In
                  values:
                  - speaker
              topologyKey: failure-domain.beta.kubernetes.io/rack
            weight: 100
...
        ports:
        - containerPort: 7472
          hostPort: 7472
          name: monitoring
          protocol: TCP
...
      hostNetwork: true
      nodeSelector:
        node-role.kubernetes.io/node: ""
```

The relevant port is definitely not in use on the nodes wight now. 
Current error message:
```
0/6 nodes are available: 3 node(s) didn't have free ports for the requested pod ports, 3 node(s) didn't match node selector.
```

Restarting the kube-scheduler did trigger the scheduling. 

Before, the scheduler had the following logs:
```
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.157467    4052 scheduler.go:447] Attempting to schedule pod: metallb-system/speaker-9f8b84ff8-22546
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.157641    4052 scheduler.go:194] Failed to schedule pod: metallb-system/speaker-9f8b84ff8-22546
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.157695    4052 factory.go:1416] Updating pod condition for metallb-system/speaker-9f8b84ff8-22546 to (PodScheduled==False)
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.159480    4052 generic_scheduler.go:1004] Node kworker-sandbox-intg-iz1-bs002 is a potential node for preemption.
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.159497    4052 generic_scheduler.go:1004] Node kworker-sandbox-intg-iz1-bs001 is a potential node for preemption.
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.159505    4052 generic_scheduler.go:1004] Node kworker-sandbox-intg-iz1-bs003 is a potential node for preemption.
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.159689    4052 scheduler.go:447] Attempting to schedule pod: metallb-system/speaker-9f8b84ff8-ngqd6
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.159800    4052 scheduler.go:194] Failed to schedule pod: metallb-system/speaker-9f8b84ff8-ngqd6
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.159843    4052 factory.go:1416] Updating pod condition for metallb-system/speaker-9f8b84ff8-ngqd6 to (PodScheduled==False)
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.161504    4052 generic_scheduler.go:1004] Node kworker-sandbox-intg-iz1-bs003 is a potential node for preemption.
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.161535    4052 generic_scheduler.go:1004] Node kworker-sandbox-intg-iz1-bs002 is a potential node for preemption.
Aug 10 19:03:28 kmaster-sandbox-intg-iz1-bs002 kube-scheduler[4052]: I0810 19:03:28.161544    4052 generic_scheduler.go:1004] Node kworker-sandbox-intg-iz1-bs001 is a potential node for preemption.
```

Changing the deployment and going through a rolling update worked fine so far. 
Deleting all pods lead to partially unschedulable pods again: 

```
> $ k get deploy -n metallb-system                                                                                                                                                                                        NAME         DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
controller   1         1         1            1           9h
speaker      4         4         4            3           9h


> $ k delete pod -n metallb-system -l component=speaker                                                                                                                                                                   pod "speaker-848c5945c9-s2jm7" deleted
pod "speaker-848c5945c9-vkw5k" deleted
pod "speaker-848c5945c9-wjg8s" deleted
pod "speaker-848c5945c9-wz4zk" deleted


> $ k get pods -n metallb-system -owide                                                                                                                                                                                   NAME                          READY     STATUS    RESTARTS   AGE       IP           NODE                             NOMINATED NODE
controller-6678f9757c-9wx28   1/1       Running   0          9h        100.64.1.5   kworker-sandbox-intg-iz1-bs001   <none>
speaker-848c5945c9-fhsmk      0/1       Pending   0          2m        <none>       <none>                           kworker-sandbox-intg-iz1-bs002
speaker-848c5945c9-fvxm4      0/1       Pending   0          2m        <none>       <none>                           kworker-sandbox-intg-iz1-bs001
speaker-848c5945c9-jqzrf      1/1       Running   0          2m        10.242.1.6   kworker-sandbox-intg-iz1-bs003   <none>
speaker-848c5945c9-xxdbs      0/1       Pending   0          2m        <none>       <none>                           <none>

> $ k get deploy -n metallb-system                                                                                                                                                                                        NAME         DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
controller   1         1         1            1           9h
speaker      4         4         4            1           9h
```

Deleting again does not change the state: 

```
> $ k delete pod -n metallb-system -l component=speaker                                                                                                                                                                   pod "speaker-848c5945c9-fhsmk" deleted
pod "speaker-848c5945c9-fvxm4" deleted
pod "speaker-848c5945c9-jqzrf" deleted
pod "speaker-848c5945c9-xxdbs" deleted

> $ k get pods -n metallb-system -owide                                                                                                                                                                                   NAME                          READY     STATUS    RESTARTS   AGE       IP           NODE                             NOMINATED NODE
controller-6678f9757c-9wx28   1/1       Running   0          9h        100.64.1.5   kworker-sandbox-intg-iz1-bs001   <none>
speaker-848c5945c9-jj7qt      0/1       Pending   0          18s       <none>       <none>                           <none>
speaker-848c5945c9-nvpsn      0/1       Pending   0          18s       <none>       <none>                           <none>
speaker-848c5945c9-phxqv      0/1       Pending   0          18s       <none>       <none>                           kworker-sandbox-intg-iz1-bs003
speaker-848c5945c9-vcf9m      0/1       Pending   0          18s       <none>       <none>                           <none>
```

Kicking the scheduler (on all masters) this time did only recover them partially:
```
> $ k get pods -n metallb-system -owide                                                                                                                                                                                   NAME                          READY     STATUS    RESTARTS   AGE       IP           NODE                             NOMINATED NODE
controller-6678f9757c-9wx28   1/1       Running   0          9h        100.64.1.5   kworker-sandbox-intg-iz1-bs001   <none>
speaker-848c5945c9-jj7qt      1/1       Running   0          6m        10.242.3.6   kworker-sandbox-intg-iz1-bs001   <none>
speaker-848c5945c9-nvpsn      1/1       Running   0          6m        10.242.2.7   kworker-sandbox-intg-iz1-bs002   <none>
speaker-848c5945c9-phxqv      0/1       Pending   0          6m        <none>       <none>                           kworker-sandbox-intg-iz1-bs003
speaker-848c5945c9-vcf9m      0/1       Pending   0          6m        <none>       <none>                           <none>
```


Do not mind that in the working state only 3 pods of 4 are scheduled. This is caused by the rack anti-affinity and only 3 worker nodes in 3 racks.

**Comment by galexrt** [comment]
I'm a colleague of @sfudeus. An observation I made is that I was initially unable to reproduce the issue (when I deleted the pods once) but when I "spammed" the delete command (`k delete pod -n metallb-system -l component=speaker`) like 5 times directly after each other, the Pods began, as described in this issue, to not recover (stayed in `Pending`).
Though I'm not sure if the "spamming" of the command is actually a way to (always; I reproduced it in the cluster 3-4 times now) reproduce the issue.

**Comment by Huang-Wei** [comment]
@sfudeus @galexrt thanks for the feedback.

I suspect it's related with "preemption" behavior which is enabled by default in 1.11. Could you try to restart kube-scheduler with flag "--featuregates=PodPriority=false", and repeat the steps to see it happens again?

And meanwhile I will look into the code.

**Comment by jpiper** [comment]
I tried with `--feature-gates=PodPriority=false` and everything seems to be working fine now, so I think your suspicions might be correct!

**Comment by Huang-Wei** [comment]
@jpiper That's great! Before we figure out a fix, please bear with disabling that as a workaround.

**Comment by sfudeus** [comment]
Deleted my pods multiple times now, have always been rescheduled when using PodPriority=false on kube-scheduler.

**Comment by jpiper** [comment]
@Huang-Wei thanks for being patient and helping us find the issue!

**Comment by sfudeus** [comment]
@Huang-Wei Thanks from my side, too. Disabling it is definitely an options which buys us sufficient time.

**Comment by Huang-Wei** [comment]
I tried to reproduce using steps provided by @sfudeus , but got slightly different symptoms - described in #67453 - and note that no matter `PodPriority` is enabled or not, it can be reproduced.

I gave a fix #67456 to fix the symptom I met. But not 100% sure that can fix this issue. I will wait for an official 1.11.x release which contains my fix, then ask you guys' favor to give a try:
- if new release helps, then it proves the two symptoms are fundamentally caused by the same bug, then this issue can be closed.
- if you still can reproduce, then it's another bug. Then I will dig into again.

**Comment by sfudeus** [comment]
@Huang-Wei Thanks for your efforts. I was not able to reproduce the issue yet with PodPriority=false in kube-scheduler on 1.11.2. Neither with sufficient nodes for the replicas nor with more replicas than nodes. 
I'm waiting for the 1.11.x featuring your fix and will then retest with PodPriority=true

**Comment by k82cn** [comment]
/close

Thanks for your contributions :)

**Comment by johnkk84** [comment]
> @Huang-Wei I just retried the scenario with 1.11.3 and `--feature-gates=PodPriority=true` on kube-scheduler and I could _**not**_ reproduce the defect. So thanks again for your efforts, I think we can close the issue.

Hi I faced with same issue on k3d after 2 years. Two things need to be set:

```
--no-lb --k3s-arg "--disable=traefik@server:0"
```

1st   - no loadbalancer for nodes
2nd  - no traefik deployment which could stole 80 and 443 ports.
